File: backend/app/models/vehicle.py
from datetime import datetime
from typing import TYPE_CHECKING, Literal, Optional

from beanie import Document, Insert, Link, Replace, before_event
from pydantic import Field
from pymongo import ASCENDING, IndexModel

from app.models.user import User

if TYPE_CHECKING:
    # Only needed for type checkers/IDEs — at runtime this is a deferred forward
    # reference, resolved in app/models/__init__.py via Vehicle.model_rebuild()
    # once QRCode has also been imported. This avoids vehicle.py <-> qr_code.py
    # trying to import each other directly at module load time.
    from app.models.qr_code import QRCode


class Vehicle(Document):
    owner: Link[User]
    vehicle_type: Literal["car", "bike"]
    vehicle_number: str
    brand: str
    model: str
    color: str
    emergency_contact: str
    qr_code_id: Optional[Link["QRCode"]] = None
    is_active: bool = True

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "vehicles"
        indexes = [
            IndexModel([("vehicle_number", ASCENDING)], unique=True),
        ]

    @before_event(Insert)
    async def _stamp_on_insert(self) -> None:
        now = datetime.utcnow()
        self.created_at = now
        self.updated_at = now

    @before_event(Replace)
    async def _stamp_on_update(self) -> None:
        self.updated_at = datetime.utcnow()

File: frontend/lib/validation/vehicle.ts
import { z } from "zod";

export const vehicleSchema = z.object({
  vehicle_type: z.enum(["car", "bike"]),
  vehicle_number: z.string().min(3, "Enter a valid vehicle number"),
  brand: z.string().min(1, "Brand is required"),
  model: z.string().min(1, "Model is required"),
  color: z.string().min(1, "Color is required"),
  emergency_contact: z
    .string()
    .min(8, "Enter a valid phone number")
    .regex(/^\+?[1-9]\d{7,14}$/, "Use international format, e.g. +919876543210"),
});
export type VehicleInput = z.infer<typeof vehicleSchema>;

export interface Vehicle extends VehicleInput {
  id: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  qr_code_id?: string | null;
}